### PR TITLE
Let `OverflowBox` be shrink-wrappable

### LINKED ISSUE #129094: `OverflowBox` always sets its own `size` as `constraints.biggest`, while users frequently needs it to "shrink wrap"
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Run the test

### Expected results

-

### Actual results

Firstly, when enableOverflowBox=false, the results are expected:

* When contentSuperLong=false, i.e. content is short, the test passes
* When contentSuperLong=true, i.e. content is too long and overflows, the test correctly prints an error saying "A RenderFlex overflowed by 9400 pixels on the bottom". That is expected, because our content is really too long.

Secondly, when enableOverflowBox=true, i.e. use OverflowBox, the results are less than optimal:

* When contentSuperLong=false, i.e. content is short, the test fails. **This is the main point the PR is going to solve**. Here, `OverflowBox` always sets its own `size` as `constraints.biggest`, while users wants it to "shrink wrap", i.e. when child is small, the OverflowBox should be as small as child.
* When contentSuperLong=true, i.e. content is too long and overflows, the OverflowBox does its job correctly.

In other words, the `enableOverflowBox=true contentSuperLong=false` needs to be fixed.

### Code Fix

Please copy-paste the following MyOverflowBox, and modify usage of OverflowBox to MyOverflowBox.

Then, as can be seen, the `enableOverflowBox=true contentSuperLong=false` case now passes test, because it correctly shrink wraps.

<details>

```dart
class MyOverflowBox extends SingleChildRenderObjectWidget {
  /// Creates a widget that lets its child overflow itself.
  const MyOverflowBox({
    super.key,
    this.alignment = Alignment.center,
    this.minWidth,
    this.maxWidth,
    this.minHeight,
    this.maxHeight,
    super.child,
  });

  /// How to align the child.
  ///
  /// The x and y values of the alignment control the horizontal and vertical
  /// alignment, respectively. An x value of -1.0 means that the left edge of
  /// the child is aligned with the left edge of the parent whereas an x value
  /// of 1.0 means that the right edge of the child is aligned with the right
  /// edge of the parent. Other values interpolate (and extrapolate) linearly.
  /// For example, a value of 0.0 means that the center of the child is aligned
  /// with the center of the parent.
  ///
  /// Defaults to [Alignment.center].
  ///
  /// See also:
  ///
  ///  * [Alignment], a class with convenient constants typically used to
  ///    specify an [AlignmentGeometry].
  ///  * [AlignmentDirectional], like [Alignment] for specifying alignments
  ///    relative to text direction.
  final AlignmentGeometry alignment;

  /// The minimum width constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  final double? minWidth;

  /// The maximum width constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  final double? maxWidth;

  /// The minimum height constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  final double? minHeight;

  /// The maximum height constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  final double? maxHeight;

  @override
  MyRenderConstrainedOverflowBox createRenderObject(BuildContext context) {
    return MyRenderConstrainedOverflowBox(
      alignment: alignment,
      minWidth: minWidth,
      maxWidth: maxWidth,
      minHeight: minHeight,
      maxHeight: maxHeight,
      textDirection: Directionality.maybeOf(context),
    );
  }

  @override
  void updateRenderObject(BuildContext context, MyRenderConstrainedOverflowBox renderObject) {
    renderObject
      ..alignment = alignment
      ..minWidth = minWidth
      ..maxWidth = maxWidth
      ..minHeight = minHeight
      ..maxHeight = maxHeight
      ..textDirection = Directionality.maybeOf(context);
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
    properties.add(DoubleProperty('minWidth', minWidth, defaultValue: null));
    properties.add(DoubleProperty('maxWidth', maxWidth, defaultValue: null));
    properties.add(DoubleProperty('minHeight', minHeight, defaultValue: null));
    properties.add(DoubleProperty('maxHeight', maxHeight, defaultValue: null));
  }
}

class MyRenderConstrainedOverflowBox extends RenderAligningShiftedBox {
  /// Creates a render object that lets its child overflow itself.
  MyRenderConstrainedOverflowBox({
    super.child,
    double? minWidth,
    double? maxWidth,
    double? minHeight,
    double? maxHeight,
    super.alignment,
    super.textDirection,
  })  : _minWidth = minWidth,
        _maxWidth = maxWidth,
        _minHeight = minHeight,
        _maxHeight = maxHeight;

  /// The minimum width constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  double? get minWidth => _minWidth;
  double? _minWidth;

  set minWidth(double? value) {
    if (_minWidth == value) {
      return;
    }
    _minWidth = value;
    markNeedsLayout();
  }

  /// The maximum width constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  double? get maxWidth => _maxWidth;
  double? _maxWidth;

  set maxWidth(double? value) {
    if (_maxWidth == value) {
      return;
    }
    _maxWidth = value;
    markNeedsLayout();
  }

  /// The minimum height constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  double? get minHeight => _minHeight;
  double? _minHeight;

  set minHeight(double? value) {
    if (_minHeight == value) {
      return;
    }
    _minHeight = value;
    markNeedsLayout();
  }

  /// The maximum height constraint to give the child. Set this to null (the
  /// default) to use the constraint from the parent instead.
  double? get maxHeight => _maxHeight;
  double? _maxHeight;

  set maxHeight(double? value) {
    if (_maxHeight == value) {
      return;
    }
    _maxHeight = value;
    markNeedsLayout();
  }

  BoxConstraints _getInnerConstraints(BoxConstraints constraints) {
    return BoxConstraints(
      minWidth: _minWidth ?? constraints.minWidth,
      maxWidth: _maxWidth ?? constraints.maxWidth,
      minHeight: _minHeight ?? constraints.minHeight,
      maxHeight: _maxHeight ?? constraints.maxHeight,
    );
  }

  @override
  // bool get sizedByParent => true;
  bool get sizedByParent => false; // NOTE MODIFIED

  // NOTE MODIFIED
  // @override
  // Size computeDryLayout(BoxConstraints constraints) {
  //   return constraints.biggest;
  // }

  @override
  void performLayout() {
    if (child != null) {
      child?.layout(_getInnerConstraints(constraints), parentUsesSize: true);
      // alignChild(); // NOTE MODIFIED
    }

    size = child != null ? constraints.constrain(child!.size) : constraints.smallest; // NOTE MODIFIED

    if (child != null) {
      alignChild();
    }

    print('hi performLayout constraints=$constraints size=$size child.size=${child?.size}');
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DoubleProperty('minWidth', minWidth, ifNull: 'use parent minWidth constraint'));
    properties.add(DoubleProperty('maxWidth', maxWidth, ifNull: 'use parent maxWidth constraint'));
    properties.add(DoubleProperty('minHeight', minHeight, ifNull: 'use parent minHeight constraint'));
    properties.add(DoubleProperty('maxHeight', maxHeight, ifNull: 'use parent maxHeight constraint'));
  }
}

```

</details>

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  for (final contentSuperLong in [false, true]) {
    for (final enableOverflowBox in [false, true]) {
      testWidgets('contentSuperLong=$contentSuperLong enableOverflowBox=$enableOverflowBox',
          (WidgetTester tester) async {
        final key = GlobalKey();

        final child = Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            SizedBox(width: 100, height: contentSuperLong ? 10000 : 100),
          ],
        );

        await tester.pumpWidget(
          MaterialApp(
            home: Stack(
              children: [
                Container(
                  key: key,
                  child: enableOverflowBox
                      ? OverflowBox(
                          maxHeight: 1000000,
                          child: child,
                        )
                      : child,
                ),
              ],
            ),
          ),
        );

        expect(tester.getBottomLeft(find.byKey(key)).dy, contentSuperLong ? 600 : 100);
      });
    }
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### PR DESCRIPTION
**Pull request by fzyzcjy**
Close https://github.com/flutter/flutter/issues/129094

I have demonstrated how this PR fixes the problem using tests in #129094. I will further add tests in this PR if the PR looks roughly acceptable :)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 1]
```diff
@@ -608,6 +610,20 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
     markNeedsLayout();
   }
 
+  /// This only affects scenario when the child does not indeed overflow.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:613]: add blank line after the first summary sentence.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 2]
```diff
@@ -608,6 +610,20 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
...
+  /// If enabled, the self size will be as small as its child when the child
+  /// does not overflow. If disabled (the default case), the self size will
+  /// always be the biggest possible size that the parent constraints allows.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:616]: "If set to true, the render object will size itself to match the size of its child within the constraints of its parent or be as small as the parent allows if no child is set. If set to false (the default), the render object will size itself to be as large as the parent allows."

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 3]
```diff
@@ -617,8 +633,10 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
...
   }
 
+  // If shrinkWrap, the size will be as small as its child when non-overflowing,
+  // thus is cannot be sizedByParent.
   @override
-  bool get sizedByParent => true;
+  bool get sizedByParent => !shrinkWrap;
 
   @override
   Size computeDryLayout(BoxConstraints constraints) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:642]: What about this implementation?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -3096,6 +3097,12 @@ class OverflowBox extends SingleChildRenderObjectWidget {
   /// default) to use the constraint from the parent instead.
   final double? maxHeight;
 
+  /// This only affects scenario when the child does not indeed overflow.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3100]: First paragraph should be a one-sentence summary.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 5]
```diff
@@ -3096,6 +3097,12 @@ class OverflowBox extends SingleChildRenderObjectWidget {
...
+  /// If enabled, the self size will be as small as its child when the child
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3101]: See comment on render object.

**Review by goderbauer — COMMENTED** [review]
Please ALWAYS add tests to your PRs. A PR is only ready for review if it has tests to prove that what is implemented actually works.

**Comment by fzyzcjy** [comment]
Thanks for the review! I will add tests soon.

**Comment by fzyzcjy** [comment]
@flutter-dashboard - I will update this PR as soon as I get back working on mobile (working on backend now)

**Comment by Piinks** [comment]
> (I got rejected too many times for PRs with full tests and green CI, just because the main idea/target does not look good... So that's why I hope to have a brief review about whether the idea is acceptable)

I would recommend discussing the proposed solution in the issue or on Discord first then, or writing a design doc and soliciting feedback there. I think this will make for a more efficient experience for both contributors and reviewers. :)

**Comment by fzyzcjy** [comment]
> I would recommend discussing the proposed solution in the issue or on Discord first then, or writing a design doc and soliciting feedback there. I think this will make for a more efficient experience for both contributors and reviewers. :)

Thanks for the suggestion! I will try that next time.

PS. I did not forget this PR, just have no time available to finish it currently, but will definitely continue later!

**Comment by Hixie** [comment]
If we do this we should probably do something like MainAxisSize or StackFit rather than calling this "shrink-wrap" (which I think we try to reserve for using intrinsic dimensions).

File: packages/flutter/lib/src/rendering/shifted_box.dart (hunk 3 above)
Comment (fzyzcjy) [reply packages/flutter/lib/src/rendering/shifted_box.dart:662]: I am not very sure about the question. I add tests testing the case now

**Comment by fzyzcjy** [comment]
Tests are added and CI is passed

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 6]
```diff
@@ -608,6 +627,23 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
     markNeedsLayout();
   }
 
+  /// The way to size the render object.
+  ///
+  /// This only affects scenario when the child does not indeed overflow.
+  /// If set to true, the render object will size itself to match the size of
+  /// its child within the constraints of its parent or be as small as the
+  /// parent allows if no child is set. If set to false (the default), the
+  /// render object will size itself to be as large as the parent allows.
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:636]: Can you please update this doc comment. The property is no longe a boolean.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 7]
```diff
@@ -617,19 +653,33 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
     );
   }
 
+  // If passthrough, the size will be as small as its child when non-overflowing,
+  // thus is cannot be sizedByParent.
   @override
-  bool get sizedByParent => true;
+  bool get sizedByParent => fit != OverflowBoxFit.passthrough;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:659]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/shifted_box.dart:659]: Also elsewhere in this PR where you compare with the enum.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 8]
```diff
@@ -617,19 +653,33 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
...
 
   @override
   Size computeDryLayout(BoxConstraints constraints) {
-    return constraints.biggest;
+    if (fit == OverflowBoxFit.passthrough) {
+      // We cannot compute dry layout, since it depends on the child
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:664]: That's not true. You can ask the child for its dry layout, no?

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/shifted_box.dart:664]: Also, consider the case where there is no child.

File: packages/flutter/test/widgets/overflow_box_test.dart [hunk 9]
```diff
@@ -30,6 +30,40 @@ void main() {
     expect(box.size, equals(const Size(100.0, 50.0)));
   });
 
+  // Adapted from https://github.com/flutter/flutter/issues/129094
```
Comment (goderbauer) [inline packages/flutter/test/widgets/overflow_box_test.dart:33]: Add a test to validate the behavior if no child is set.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 10]
```diff
@@ -515,6 +515,23 @@ class RenderPositionedBox extends RenderAligningShiftedBox {
...
+enum OverflowBoxFit {
+  /// The widget will size itself to be as large as the parent allows.
+  max,
+
+  /// The widget will follow the child's size.
+  ///
+  /// More specifically, the render object will size itself to match the size of
+  /// its child within the constraints of its parent or be as small as the
+  /// parent allows if no child is set.
+  passthrough,
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:532]: This name isn't great. Other suggestions: `min` (as in be as small as possible without clipping the child), or `deferToChild` to be more explicit.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 11]
```diff
@@ -515,6 +515,23 @@ class RenderPositionedBox extends RenderAligningShiftedBox {
   }
 }
 
+/// How much space should be occupied by the [OverflowBox].
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/shifted_box.dart:518]: add: ..if there is no overflow?

**Comment by fzyzcjy** [comment]
Thank you for the review! I am busy with other subjects so will come back when having time!

**Comment by fzyzcjy** [comment]
CI is green now

image

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 12]
```diff
@@ -511,6 +511,20 @@ class RenderPositionedBox extends RenderAligningShiftedBox {
...
+/// How much space should be occupied by the [OverflowBox] if there is no
+/// overflow.
+enum OverflowBoxFit {
+  /// The widget will size itself to be as large as the parent allows.
+  max,
+
+  /// The widget will follow the child's size.
+  ///
+  /// More specifically, the render object will size itself to match the size of
+  /// its child within the constraints of its parent or be as small as the
```
Comment (cbracken) [inline packages/flutter/lib/src/rendering/shifted_box.dart:523]: Super minor optional clarification/simplification:
```suggestion
  /// its child within the constraints of its parent, or as small as the
```

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 13]
```diff
@@ -604,6 +620,24 @@ class RenderConstrainedOverflowBox extends RenderAligningShiftedBox {
     markNeedsLayout();
   }
 
+  /// The way to size the render object.
+  ///
+  /// This only affects scenario when the child does not indeed overflow.
+  /// If set to [OverflowBoxFit.deferToChild], the render object will size
+  /// itself to match the size of its child within the constraints of its
+  /// parent or be as small as the parent allows if no child is set.
```
Comment (cbracken) [inline packages/flutter/lib/src/rendering/shifted_box.dart:628]:
```suggestion
  /// parent, or as small as the parent allows if no child is set.
```

**Review by cbracken — APPROVED** [review]
lgtm!

**Comment by goderbauer** [comment]
@cbracken The CLA bot seems to be unhappy about the commits you have co-authored in this PR?! Is that a config issue on your end?

**Comment by cbracken** [comment]
I don't think so; I've never had the CLA bot shout about suggested commits in recent memory. Seems like potentially a bug in the CLA bot? The is implicit for all Googlers, but I've also signed it manually (years ago, back in 2013) just in case.

**Comment by cbracken** [comment]
I think I have a guess as to what happened. I just tried to manually re-sign the CLA and it wants you to log in with a GAIA account with that email address. Until recently I had my personal custom-domain email hosted via Google Workspaces but I migrated to another solution and deleted the dasher domain last week. I wonder if the CLA bot requires an email account managed by Google.

I can see about trying to restore the dasher domain this week, but given that we know I'm an employee, I think we're safe to merge regardless of CLA bot status.

**Comment by goderbauer** [comment]
OK, thanks @cbracken. I am going to do a rebase to kick the google testing job and if the CLA complains again override it.

**Comment by goderbauer** [comment]
@fzyzcjy Can you resolve the conflicts and then we should be able to get this in.

**Comment by fzyzcjy** [comment]
Sure!

**Pull request merged**